scripts/xtensa-build-zephyr.py

```diff
@@ -0,0 +1,382 @@
+
+def check_west_installation():
+	west_path = shutil.which("west")
+	if not west_path:
+		raise FileNotFoundError("Install west and a west toolchain,"
+			"https://docs.zephyrproject.org/latest/getting_started/index.html")
+	print(f"Found west: {west_path}")
+
+def find_west_workspace():
+	"""[summary] Scans for west workspaces using west topdir command by checking script execution
+	WEST_TOP and SOF_TOP directories.
+
+	:return: [description] Returns path when west workspace had been found, otherwise None
+	:rtype: [type] bool, pathlib.Path
+	"""
+	global WEST_TOP, SOF_TOP
+	paths = [ pathlib.Path(os.getcwd()), SOF_TOP, WEST_TOP]
+	for path in paths:
+		if path.is_dir():
+			result = execute_command("west topdir", stderr=subprocess.DEVNULL,
+				stdout=subprocess.DEVNULL, check=False, cwd=path)
+		if result.returncode == 0:
+			return path
+	return None
+
+def create_zephyr_directory():
+	global WEST_TOP
+	WEST_TOP.mkdir(mode=511, parents=False, exist_ok=False)
+	WEST_TOP = WEST_TOP.resolve(strict=True)
+
+def create_zephyr_sof_symlink():
+	global WEST_TOP, SOF_TOP
+	audio_modules_dir = pathlib.Path(WEST_TOP, "modules", "audio")
+	audio_modules_dir.mkdir(mode=511, parents=True, exist_ok=True)
+	sof_symlink = pathlib.Path(audio_modules_dir, "sof")
+	# Symlinks creation requires administrative privileges in Windows or special user rights
+	try:
+		if not sof_symlink.is_symlink():
+			sof_symlink.symlink_to(SOF_TOP, target_is_directory=True)
+	except:
+		print(f"Failed to create symbolic link: {sof_symlink} to {SOF_TOP}."
+			"\nIf you run script on Windows run it with administrative privileges or\n"
+			"grant user symlink creation rights -"
+			"see: https://docs.microsoft.com/en-us/windows/security/threat-protection/"
+			"security-policy-settings/create-symbolic-links")
+		raise
+
+def west_init_update():
+	global WEST_TOP, SOF_TOP
+	zephyr_dir = pathlib.Path(WEST_TOP, "zephyr")
+	execute_command(f"git clone {args.url} {zephyr_dir}", check=True, timeout=1200)
+	execute_command(f"git fetch origin {args.z}", check=True, timeout=300, cwd=zephyr_dir)
+	execute_command(f"git checkout FETCH_HEAD", check=True, cwd=zephyr_dir)
+	execute_command(f"git -C {zephyr_dir} --no-pager log --oneline --graph --decorate"
+		" --max-count=20", check=True)
+	execute_command(f"west init -l {zephyr_dir}", check=True)
+	create_zephyr_sof_symlink()
+	execute_command("west update zephyr hal_xtensa", check=True, timeout=300, cwd=WEST_TOP)
+
+def build_platforms():
+	global WEST_TOP, SOF_TOP
+	# double check that west workspace is initialized
+	execute_command("west topdir", check=True, stderr=subprocess.DEVNULL, stdout=subprocess.DEVNULL,
+		cwd=WEST_TOP)
+	print(f"SOF_TOP={SOF_TOP}")
+	print(f"WEST_TOP={WEST_TOP}")
+	global STAGING_DIR
+	STAGING_DIR = pathlib.Path(WEST_TOP, "build-sof-staging", "sof")
+	# smex does not use 'install -D'
+	STAGING_DIR.mkdir(mode=511, parents=True, exist_ok=True)
+	for platform in args.platforms:
```

Comment:
To save a ton of whitespace you could move all this to a new, `build_one_platform()` function (I always wanted to do this in the shell script).

Then the `build_platforms()` function can probably be inlined and disappear.

Reply:
We can do this later but it will cause a lot of whitespace churn in git history. No big deal.

Reply:
Lets get it stable first.